src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/JenkinsHash.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Text;
+
+namespace Orleans.Storage
+{
+    // Based on the version in http://home.comcast.net/~bretm/hash/7.html, which is based on that
+    // in http://burtleburtle.net/bob/hash/evahash.html.
+    // Note that we only use the version that takes three ulongs, which was written by the Orleans team.
+    // implementation restored from Orleans v3.7.2: https://github.com/dotnet/orleans/blob/b24e446abfd883f0e4ed614f5267eaa3331548dc/src/Orleans.Core.Abstractions/IDs/JenkinsHash.cs
+    internal static class JenkinsHash
+    {
+        private static void Mix(ref uint a, ref uint b, ref uint c)
+        {
+            a -= b; a -= c; a ^= (c >> 13);
```

Comment:
As `a`, `b`, and `c` are `ref`s, each operation must access the (stack) memory.
When the operations are solely done on locals, then registers can be used.

Cf. [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhhNySJigYBZAQgAUsAGYMArgIB2GBtjQNVG7buAGjmnQzABKLj07UezvQzgBeBsADcrj9d9sBgA9T0UwBgA+SIZyOhtvexcvN08wX2BUvQyQsKCAHnyGAA4EpJcI/2xfSs8fa1yGRUzo2PjEp2S/OsCs9NdQpojW8lIyzuTM/36pz2qUwcUCoslxroaqmqz6iMWWmIBWNa6g/3rTtN7F4Zj28udZgJTNhbyGQtiABmPk2uyNnoNPZRGLkI4dZwAX3K5X4wlE4mIkmkcgU5GUMDUll02H0hkxxiswDM+KxJmstnKjnW2Ncc2q9x4tMydR8jO4tIiaXS7IYvIuXl60yuYRubR+LkeMyy80yi2WJQlzj+8z+O0azRB4ohJ22QsuA1FWtGSp4UpyLzlbw+qx1vxlWzOW2BrXB/L13SeQWuWruE0lfQtcxy8veK2+doqDoBgqBYX2sTd/uc/KCwd5xIYPV5YC5AXK0OokKAA)

Reply:
Out of curiosity I ran a benchmark.

```
| Method     | Mean      | Error     | StdDev    | Ratio | RatioSD |
|----------- |----------:|----------:|----------:|------:|--------:|
| Default    | 27.784 ns | 0.5689 ns | 0.5322 ns |  1.00 |    0.03 |
| Suggestion |  6.043 ns | 0.1590 ns | 0.1633 ns |  0.22 |    0.01 |
```
So roughly 4.5x faster (in the isolated benchmark).


<details>
  <summary>code</summary>

```c#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Bench>();

public class Bench
{
    private uint _a = 42;
    private uint _b = 43;
    private uint _c = 44;

    [Benchmark(Baseline = true)]
    public void Default() => Mix(ref _a, ref _b, ref _c);

    [Benchmark]
    public void Suggestion() => Mix1(ref _a, ref _b, ref _c);

    private static void Mix(ref uint a, ref uint b, ref uint c)
    {
        a -= b; a -= c; a ^= (c >> 13);
        b -= c; b -= a; b ^= (a << 8);
        c -= a; c -= b; c ^= (b >> 13);
        a -= b; a -= c; a ^= (c >> 12);
        b -= c; b -= a; b ^= (a << 16);
        c -= a; c -= b; c ^= (b >> 5);
        a -= b; a -= c; a ^= (c >> 3);
        b -= c; b -= a; b ^= (a << 10);
        c -= a; c -= b; c ^= (b >> 15);
    }

    private static void Mix1(ref uint aa, ref uint bb, ref uint cc)
    {
        uint a = aa;
        uint b = bb;
        uint c = cc;

        a -= b; a -= c; a ^= (c >> 13);
        b -= c; b -= a; b ^= (a << 8);
        c -= a; c -= b; c ^= (b >> 13);
        a -= b; a -= c; a ^= (c >> 12);
        b -= c; b -= a; b ^= (a << 16);
        c -= a; c -= b; c ^= (b >> 5);
        a -= b; a -= c; a ^= (c >> 3);
        b -= c; b -= a; b ^= (a << 10);
        c -= a; c -= b; c ^= (b >> 15);

        aa = a;
        bb = b;
        cc = c;
    }
}
```
</details>

Reply:
The idea was to keep the implementation of the `JenkinsHash` exactly the same it was in 3.7.2, without introducing potentially breaking changes.

It's possible to improve it but only with a proper coverage by tests.

Reply:
That's true, but think in a different way:

For each operation memory needs to be accessed, so if in L1 read from there, otherwise go the cache hierarchy up to RAM if not found. So we have potential for cache trashing, which can't be shown in micro-benchmarks.

When memory access is minimized, the chance for cache trashing is minimized too.
Here the code change required is trivial, so I'd go with the suggestion -- the algorithm is untouched, just at entry and exit of the method the value are read to local / written back.

Reply:
OK, the change done